API deprecation policy (#205)
redpanda-data/cloud-docs: modules/manage/pages/api/cloud-api-deprecation-policy.adoc

= Redpanda Cloud API Deprecation Policy
:description: Deprecation and versioning policy for Redpanda Cloud API.
:page-api: cloud

> This document outlines the Cloud API deprecation and versioning policy to ensure a smooth transition for Redpanda Cloud users. To maintain a reliable and robust API that supports your integration needs, it is sometimes necessary to deprecate older versions or features of the Redpanda Cloud API as Redpanda introduces improvements, addresses security concerns, and updates the technology stack. Redpanda adheres to this policy to maintain a balance between innovation and stability, ensuring the best possible experience for customers and partners. 

Reviewer: ```suggestion
This document outlines the Cloud API deprecation and versioning policy. As Redpanda introduces improvements, addresses security concerns, and updates the technology stack, it is sometimes necessary to deprecate older versions or features of the Redpanda Cloud API. Redpanda adheres to this policy to maintain a balance between innovation and stability, ensuring the best possible experience for customers and partners. 
```

> For questions and concerns regarding the policy, contact https://support.redpanda.com/hc/en-us/requests/new[Redpanda Support^]. 

Reviewer: ```suggestion
For questions and concerns regarding this policy, contact https://support.redpanda.com/hc/en-us/requests/new[Redpanda Support^]. 
```

Reviewer 2: Thanks for the tag, do we know the expected questions, or any potential answers?  I suspect product would need to provide that.  Unless there is a technical challenge, I would prefer customers go to their CSMs and community users use the community.

Reviewer 3: Agreed with Shane. This should be addressed via CSMs. Especially for our bigger customers. That being said, what kind of questions are you expecting?

== Definitions

> Deprecation is the process of phasing out an API version or specific features within an API. When deprecated, these versions or features no longer receive updates, support, or guarantees of availability, and are not usable at the end of the deprecation period.

Reviewer: ```suggestion
Deprecation is the process of phasing out an API version or specific features within an API. At the end of the deprecation period, these versions or features no longer receive updates, support, or guarantees of availability, and they are no longer usable.
```

The Cloud API availability phases are defined as follows: 

> * Limited availability (LA): The API is available to select customers in a subset of regions, providers, or networks. Recommended for production usage.
> * Generally available (GA): The API is stable. Protocol breaking changes are not introduced on short notice. Recommended for production usage. 

Reviewer: ```suggestion
* Generally available (GA): The API is stable. Protocol breaking changes are not introduced on short notice. Recommended for production usage. 
* Limited availability (LA): The API is available to select customers in a subset of regions, providers, or networks. Recommended for production usage.
```

Reviewer 4: @kbatuigas "breaking changes are not introduced on short notice." Can't this simply be "...and do not include breaking changes"? Unless you specifically define what you mean here by "short notice", this is not very clear to users. 

Also, is LA really "recommended" for production use, or is it available (not necessarily recommended) for production?

Author: > @kbatuigas "breaking changes are not introduced on short notice." Can't this simply be "The API is stable and does not include protocol-breaking changes"? Unless you specifically define what you mean here by "short notice", this is not very clear to users.
> 
> Also, is LA really "recommended" for production use, or is it available (not necessarily recommended) for production?

@towfiqa could you confirm re: the question about LA?

Reviewer: Could we just remove LA entirely here?

* Deprecated: The API version or method is no longer under active development. Existing workflows and integrations using the API continue to function, but new use cases should be built against the new version. The deprecated API version or method is retired on the retirement date. 
* Retired: The API version or method is no longer supported nor available.

== Deprecation notice and timelines

. Redpanda provides a formal deprecation notice to users when an API version or API method is scheduled for deprecation. This notice includes:
+
--

> * The specific API version or method to be deprecated.

Reviewer: I don't think this section should use a numbered list: these aren't steps. I'd also merge the line "Users have a transition period..." into the End-of-Support date item. And I'd merge the section that starts with "Starting from the retirement date..." into the Retirement date item.

Reviewer 4: absolutely +1

* The reason for deprecation.
* The deprecation timeline, including key dates:
** Announcement date: The date when the deprecation is officially announced. The API version or method remains available through the end-of-support date, until the retirement date.
** End-of-Support date: The last date the deprecated API version or call is supported. The API is available, but is not under active development and no longer receives support. This date is six months after the announcement date.
** Retirement date: The date when the deprecated API or feature is removed and no longer available. This date is 12 months after the announcement date.
--
. Users have a transition period between the announcement date and the retirement date to migrate to new API versions or methods.
. Starting from the retirement date, calls to the deprecated API version or method may:
+
--
* Be rejected with appropriate error codes.
* Result in degraded performance.
* Result in unexpected behavior.
--

> To avoid disruption, you are encouraged to monitor usage of deprecated methods and update integrations promptly.

Reviewer: ```suggestion
TIP: To avoid disruption, you are encouraged to monitor usage of deprecated methods and update integrations promptly.
```

=== Exceptions to standard timeline

> In rare cases, the deprecation process may need to be accelerated due to:
> 
> * Critical security vulnerabilities.
> * Compliance requirements.

Reviewer: ```suggestion
In rare cases, the deprecation process may need to be accelerated. For example:

* Critical security vulnerabilities
* Compliance requirements
```

* External infrastructure changes outside Redpanda's control.

== Notification channels

Redpanda communicates deprecation notices through the following:

> * A `Sunset` HTTP header in the response 

Reviewer: ```suggestion
* `Sunset` HTTP header in the response 
```

> * `deprecated: true` field in the API specification

Reviewer: ```suggestion
* A `deprecated: true` field in the API specification
```

* Updates to xref:manage:api/index.adoc[Cloud API documentation]
* xref:get-started:whats-new-cloud.adoc[What’s New in Redpanda Cloud]
* Announcements in internal customer Slack channels

> * #cloudapi Slack channel at redpandacommunity.slack.com

Reviewer: You could merge this with the line above: Announcements in customer Slack channels and in the #cloudapi Community Slack channel

* Emails to “Sign up for updates” recipients (sign up on the xref:api:ROOT:cloud-controlplane-api.adoc[Control Plane API] and xref:api:ROOT:cloud-dataplane-api.adoc[Data Plane APIs] reference documentation)

== Versioning policy

> Redpanda Cloud API versions are denoted in the API URL with the following format: `vX`, for example, `v1`. The version number is incremented when there are breaking changes to the API.

Reviewer: ```suggestion
Redpanda Cloud API versions are denoted in the API URL with the following format: `vX`. For example: `v1`. The version number is incremented when there are breaking changes to the API.
```

Reviewer 4: +1

> The Control Plane API and Data Plane API may be versioned independently of each other, and individual components such as RBAC (IAM) consisting of Control Plane Users, Control Plane Service Accounts, and Control Plane Role Bindings may also have a different version number than the base API. 

Reviewer: ```suggestion
The Control Plane API and Data Plane API may be versioned independently of each other, and individual components may have a different version number than the base API, such as RBAC (IAM).
```

= Redpanda Cloud API Deprecation Policy
:description: Deprecation and versioning policy for Redpanda Cloud API.
:page-api: cloud

This document outlines the Cloud API deprecation and versioning policy to ensure a smooth transition for Redpanda Cloud users. To maintain a reliable and robust API that supports your integration needs, it is sometimes necessary to deprecate older versions or features of the Redpanda Cloud API as Redpanda introduces improvements, addresses security concerns, and updates the technology stack. Redpanda adheres to this policy to maintain a balance between innovation and stability, ensuring the best possible experience for customers and partners. 

For questions and concerns regarding the policy, contact https://support.redpanda.com/hc/en-us/requests/new[Redpanda Support^]. 

== Definition

Deprecation is the process of phasing out an API version or specific features within an API. When deprecated, these versions or features no longer receive updates, support, or guarantees of availability, and are not usable at the end of the deprecation period.

> === Availability

Reviewer 3: I think we can remove this heading entirely. And maybe just bold the word being defined.

To avoid disruption, you are encouraged to monitor usage of deprecated methods and update integrations promptly.

=== Exceptions to standard timeline

In rare cases, the deprecation process may need to be accelerated due to:

* Critical security vulnerabilities.
* Compliance requirements.
* External infrastructure changes outside Redpanda's control.

== Notification channels

Redpanda communicates deprecation notices through the following:

* A `Sunset` HTTP header in the response 
* `deprecated: true` field in the API specification
* Updates to xref:manage:api/index.adoc[Cloud API documentation]
* xref:get-started:whats-new-cloud.adoc[What’s New in Redpanda Cloud]
* Announcements in internal customer Slack channels
* #cloudapi Slack channel at redpandacommunity.slack.com
* Emails to “Sign up for updates” recipients (sign up on the xref:api:ROOT:cloud-controlplane-api.adoc[Control Plane API] and xref:api:ROOT:cloud-dataplane-api.adoc[Data Plane APIs] reference documentation)

== Versioning policy

Redpanda Cloud API versions are denoted in the API URL with the following format: `vX`, for example, `v1`. The version number is incremented when there are breaking changes to the API.

> The Control Plane API and Data Plane API may be versioned independently of each other, and individual components such as RBAC (IAM) consisting of Control Plane Users, Control Plane Service Accounts, and Control Plane Role Bindings may also have a different version number than the base API. 

Author: @birdayz @sago2k8 @yougotashovel Is this correct?

cc @towfiqa
